### ISSUE TITLE
Корректна ли работа пользовательской подборки?

### ISSUE THREAD

**Issue by kbochenina** [avatar-project/public-issues#77]
(1) Создаю подборку, теги Стажировка - Internship - Вакансия - Работа - Art - Science - Лекция - Кинопоказ
(2) Создаю подборку, тэги искусство творчество семинар лекция

Выдается 1 выдача - вакансия от музея оптики, и в том и в другом случае. Выделен тег Вакансия (в случае № 2 это даже фактически неверно)

**Comment by hainamantik** [comment]
Какой стенд?

**Comment by hainamantik** [comment]
Стенд тест1. На бэке в подборках 101 выдача и 29 выдач соответственно.
Возможно на фронте всё ещё не пофиксился баг с отображанием контента не из той подборки.
@kbochenina если перезапустить приложение и снова зайти в одну из этих подборок, там всё ещё одна выдача?

**Comment by kbochenina** [comment]
@hainamantik в этих, о которых написано в данном тикете, обновилось после повторного входа (про dev речь).

Но сейчас вообще оч странное поведение:

Создаю подборку с тэгами футбол, функан, футуризм, захожу - там единственная выдача про
 
![photo_2021-02-20_11-50-16](https://user-images.githubusercontent.com/5405987/108589909-dbbc9500-7371-11eb-8b6c-69b716a37a1e.jpg)

После перезагрузки приложения туда нагрузилось 25 выдач с какими то произвольными тэгами, от марсохода до Kotlin

Хотя если окно редактирования открыть - тэги по прежнему футбол, функан, футуризм

**Comment by kbochenina** [comment]
По моему там что-то путается при выводе информации между подборками

**Comment by hainamantik** [comment]
@danwhale 
> Создаю подборку с тэгами футбол, функан, футуризм, захожу - там единственная выдача про
```
tags: [
    'футбол',
    'функан',
    'футуризм'
],
similar_tags: [
    'чемпионат',
    'спорт',
    'матан',
    'science',
    'digital',
    'лекция'
],
```
похожие теги вообще не в тему.

@kbochenina ассет добавился в подборку на основании похожих тегов, почему они именно такие - не знаю

**Comment by kbochenina** [comment]
(1) Честно, было бы конечно идеально похожие тэги в явном виде догружать в окно создания подборки, чтобы пользователь мог вручную редактировать  список 

similar_tags: [
    'чемпионат',
    'спорт',
    'матан',
    'science',
    'digital',
    'лекция'
],

и у него бы в подборке было только то что соответствует тем тэгам которые он хочет. С явным подсвечиванием этих тэгов. (Потому что информационный фильтр - "хочу найти", он так и работает)

(2) Если это невозможно сделать, то тогда хоть предупреждать в окне создания, изменив надпись "Добавьте от 3 до 10 тегов (Аватар будет подбирать информацию по выбранным и похожим тегам")

Призываются @hainamantik @kapc3d @danwhale

**Comment by kapc3d** [comment]
надо как то логику тогда продумать по похожим, потому что щас поиск чисто под добавляемые теги. можно конечно пассивный вывод делать похожишь - тег добавил - список похожих обновился. ручками редактировать еще и похожие - слишком сложно для пользователя, этим не будут пользоваться, всегда оставлять по дефолту.

но для визуализации похожих нужно чтобы нам их отдавали.

**Comment by kbochenina** [comment]
@hainamantik можете похожие отдавать чтоб мы их могли показывать в окне создания подборки?

> ручками редактировать еще и похожие - слишком сложно для пользователя, этим не будут пользоваться, всегда оставлять по дефолту.

Согласна.

Если это сделать то работать будет примерно так:
- выбрал футбол, тебе серенькими (неактивными) подставился под зелеными тэгами в окне создания подборки еще чемпионат и спорт, убрал футбол, убрались чемпионат и спорт, выбрал функан, добавился еще матан, ну  и тд

И тогда сама фильтрация останется такой же.

**Comment by danwhale** [comment]
С этим нужно подумать, на первый взгляд с нашей стороны для этого все возможности есть, с другой надо проработать все кейсы. Ну и плюс на последний день выносим фичу как всегда. Я пообедаю и точно смогу сказать :)

**Comment by kbochenina** [comment]
@danwhale я второй вариант на крайний случай также предложила - текстовое пояснение. 

Про последний день - тут вижу проблему в том, что НЕ документировалась реализация (что подбор не строго по указанным тегам, а по похожим). И соответственно не было раньше возможности оценить это с т.зр. как подать вот эту конкретную реализацию фильтра.

**Comment by kbochenina** [comment]
Но вообще, конечно гораздо лучше если бы пользователь видел список тегов, по которым идет подбор. Тогда все понятно - вот твой фильтр, ты в него заходишь, и на выдачах видишь отметки тегов (выбранных + похожих). 

А сейчас тк еще не отлажена система "похожих", это выглядит так:
- указал свои теги
- получил подборку, в которой на выдачах другие теги

**Comment by danwhale** [comment]
В целом можем выдать ручку с поиском похожих тегов наружу и убрать из внутренней рекомендашки (ассоциированные теги), будут тока те, которые добавит юзер осознанно - здесь падает работа на фронт @kapc3d 

либо просто их выпилить и забыть, но пользовательские подборки будут совсем пустые, мы самообучающуюся штуку с поиском похожих тегов не успели встащить по времени, потому получаем результат itmonews 

Идея была в том, что подборки пустые, а такая фича позволяет нам меньше давать нагрузку на юзера и приносить похожий по смыслу (статистически) контент.

**Comment by kbochenina** [comment]
Да нет, выпилить сейчас вообще не вариант (будут пустые).

> надо как то логику тогда продумать по похожим, потому что щас поиск чисто под добавляемые теги. можно конечно пассивный вывод делать похожишь - тег добавил - список похожих обновился. ручками редактировать еще и похожие - слишком сложно для пользователя, этим не будут пользоваться, всегда оставлять по дефолту.
> 
> но для визуализации похожих нужно чтобы нам их отдавали.

Вот это нормально, что @kapc3d предлагал. По похожим ведется поиск - просто пользователь об этом ЗНАЕТ.

**Comment by kapc3d** [comment]
я могу ошибаться в деталях текущей реализации, но
механизм предлагается такой:

> при каждом добавлении тега мы шлем вам список добавленных тего, вы в ответ шлете список похожишь, мы их дорисовываем ниже некликабельными

@danwhale Верно?

**Comment by danwhale** [comment]
> я могу ошибаться в деталях текущей реализации, но
> механизм предлагается такой:
> 
> > при каждом добавлении тега мы шлем вам список добавленных тего, вы в ответ шлете список похожишь, мы их дорисовываем ниже некликабельными
> 
> @danwhale Верно?

верно, я сейчас проверю, возможно мы даже без изменения кода бэкэнда это реализуем.

**Comment by danwhale** [comment]
Есть у нас роут `feed/groups/tag_recommend` -- туда можно через пробел подать текстом выбранные  теги, оно вернёт список тегов (как текуших) так и близких к ним, если фронт согласен на своей стороне вычитать из пришедшего множеста выбранные теги, работать будет корректно

**Comment by kapc3d** [comment]
блин, на каждом запросе перебирать все это, ну такое... выглядит как дикий костыль
надо с ребятами посоветоваться

**Comment by danwhale** [comment]
> блин, на каждом запросе перебирать все это, ну такое... выглядит как дикий костыль
> надо с ребятами посоветоваться

С нашей стороны это тоже костыль на самом деле, тк там есть довольно дорогая промежуточная операция. @hainamantik в принципе мы можем вытащить ещё один роут который:
- принимает на вход массив тегов
- возвращает список похожих тегов по аналогии с формированием `similar tags` у нас под капотом при создании подборки

?

**Comment by hainamantik** [comment]
> > блин, на каждом запросе перебирать все это, ну такое... выглядит как дикий костыль
> > надо с ребятами посоветоваться
> 
> С нашей стороны это тоже костыль на самом деле, тк там есть довольно дорогая промежуточная операция. @hainamantik в принципе мы можем вытащить ещё один роут который:
> 
>     * принимает на вход массив тегов
> 
>     * возвращает список похожих тегов по аналогии с формированием `similar tags` у нас под капотом при создании подборки
> 
> 
> ?

если сильно нужно, то сделаю

**Comment by danwhale** [comment]
@kapc3d  Андрей, решите тогда, в целом вычитание маленьких сетов на каждое обновление не такая тяжелая операция
